From 3819330da9179904c11c728e599a542f373b75f4 Mon Sep 17 00:00:00 2001
From: Geoff Thorpe <Geoff.Thorpe@freescale.com>
Date: Wed, 11 Nov 2009 21:26:56 -0500
Subject: [PATCH] pme2: fix a race-condition in pme_ctx_ctrl_update_flow()

Signed-off-by: Geoff Thorpe <Geoff.Thorpe@freescale.com>
[Kevin: Original patch taken from Freescale p4080 SDK 2.0.1 ISO image.
Apply cleanly to kernel 2.6.27]
Integrated-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/match/pme2_high.c |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/match/pme2_high.c b/drivers/match/pme2_high.c
index 729ef21..61a25c8 100644
--- a/drivers/match/pme2_high.c
+++ b/drivers/match/pme2_high.c
@@ -596,12 +596,17 @@ int pme_ctx_ctrl_update_flow(struct pme_ctx *ctx, u32 flags,
 			flags &= ~PME_CMD_FCW_RES;
 	}
 	/* allocate residue memory if it is being added */
-	if ((flags & PME_CMD_FCW_RES) && params->ren && !ctx->hw_residue) {
-		ctx->hw_residue = pme_hw_residue_new();
+	if ((flags & PME_CMD_FCW_RES) && params->ren) {
+		spin_lock_irq(&ctx->lock);
 		if (!ctx->hw_residue) {
-			pme_hw_flow_free(token->internal_flow_ptr);
-			return -ENOMEM;
+			ctx->hw_residue = pme_hw_residue_new();
+			if (!ctx->hw_residue) {
+				spin_unlock_irq(&ctx->lock);
+				pme_hw_flow_free(token->internal_flow_ptr);
+				return -ENOMEM;
+			}
 		}
+		spin_unlock_irq(&ctx->lock);
 	}
 	/* enqueue the FCW command to PME */
 	memcpy(token->internal_flow_ptr, params, sizeof(struct pme_flow));
-- 
1.6.0.4

